api-update/common/src/main/scala/uk/gov/nationalarchives/api/update/common/ApiUpdate.scala

```diff
@@ -0,0 +1,41 @@
+package uk.gov.nationalarchives.api.update.common
+
+import sangria.ast.Document
+import uk.gov.nationalarchives.api.update.common.exceptions.AuthorisationException
+import uk.gov.nationalarchives.api.update.common.exceptions.{AuthorisationException, GraphQlException}
+import uk.gov.nationalarchives.tdr.error.NotAuthorisedError
+import uk.gov.nationalarchives.tdr.keycloak.KeycloakUtils
+import uk.gov.nationalarchives.tdr.{GraphQLClient, GraphQlResponse}
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.language.postfixOps
+import scala.util.{Failure, Success, Try}
+
+class ApiUpdate()(implicit val executionContext: ExecutionContext) {
+
+  def send[D, V](keycloakUtils: KeycloakUtils, client: GraphQLClient[D, V], document: Document, variables: V): D = {
+    val queryResult: Future[GraphQlResponse[D]] = for {
+      token <- keycloakUtils.serviceAccountToken(sys.env("CLIENT_ID"), sys.env("CLIENT_SECRET"))
+      result <- client.getResult(token, document, Option(variables))
+    } yield result
+
+    val result: Future[D] = queryResult.map(result => {
+      result.errors match {
+        case Nil => result.data.get
+        case List(authError: NotAuthorisedError) => throw new AuthorisationException(authError.message)
+        case errors => throw new GraphQlException(errors)
+      }
+    })
+
+    val output: Try[D] = Await.ready(result, 20 seconds).value.get
```

Comment:
Is it possible to do the `Await` in the top level of the Lambda function and return a Future from this function?

I don't expect we'll need to change this module very often or add more code, but it sets a good example if we only resolve futures when absolutely necessary (e.g. to wrap up the Lambda execution, rather than at the point where we make the API call)

Reply:
Yeah that makes, sense. It's also possible that we could use this project for code which will run on ECS or elsewhere that can be a bit more Future friendly. I've changed it anyway.